File: packages/sdk-python/openpassport/canonical.py
"""
Deterministic JSON serialization for signing.
Must produce byte-identical output to the JavaScript implementation.

Rules:
1. Remove excluded keys (default: "signature")
2. Sort remaining keys alphabetically by Unicode code point
3. Serialize with no whitespace, using separators=(",", ":")
4. Nested objects are also sorted recursively
"""

from __future__ import annotations

import json
from typing import Any, Optional, List


def canonicalize(obj: dict[str, Any], exclude_keys: Optional[List[str]] = None) -> str:
    if exclude_keys is None:
        exclude_keys = ["signature"]
    cleaned = _sort_keys(obj, exclude_keys)
    return json.dumps(cleaned, separators=(",", ":"), ensure_ascii=False)


def _sort_keys(value: Any, exclude_keys: list[str]) -> Any:
    if value is None or isinstance(value, (bool, int, float, str)):
        return value
    if isinstance(value, list):
        return [_sort_keys(item, []) for item in value]
    if isinstance(value, dict):
        return {
            k: _sort_keys(v, [])
            for k, v in sorted(value.items())
            if k not in exclude_keys
        }
    return value

File: packages/spec/src/canonical.ts
/**
 * Deterministic JSON serialization for signing.
 * Produces identical bytes across JS and Python implementations.
 *
 * Rules:
 * 1. Remove excluded keys (default: "signature")
 * 2. Sort remaining keys alphabetically by Unicode code point
 * 3. Serialize with no whitespace
 * 4. Nested objects are also sorted recursively
 */
export function canonicalize(
  obj: Record<string, unknown>,
  excludeKeys: string[] = ["signature"],
): string {
  return JSON.stringify(sortKeys(obj, excludeKeys));
}

function sortKeys(
  value: unknown,
  excludeKeys: string[],
): unknown {
  if (value === null || typeof value !== "object") {
    return value;
  }

  if (Array.isArray(value)) {
    return value.map((item) => sortKeys(item, excludeKeys));
  }

  const obj = value as Record<string, unknown>;
  const sorted: Record<string, unknown> = {};
  const keys = Object.keys(obj)
    .filter((k) => !excludeKeys.includes(k))
    .sort();

  for (const key of keys) {
    sorted[key] = sortKeys(obj[key], []);
  }

  return sorted;
}
